chips/nrf52/src/chip.rs

```diff
@@ -57,61 +135,8 @@ impl kernel::Chip for NRF52 {
-                                    "nRF 802.15.4 and BLE radios cannot be simultaneously enabled!"
-                                ),
-                            }
-                        }
-                        peripheral_interrupts::RNG => nrf5x::trng::TRNG.handle_interrupt(),
-                        peripheral_interrupts::RTC1 => nrf5x::rtc::RTC.handle_interrupt(),
-                        peripheral_interrupts::TEMP => nrf5x::temperature::TEMP.handle_interrupt(),
-                        peripheral_interrupts::TIMER0 => nrf5x::timer::TIMER0.handle_interrupt(),
-                        peripheral_interrupts::TIMER1 => nrf5x::timer::ALARM1.handle_interrupt(),
-                        peripheral_interrupts::TIMER2 => nrf5x::timer::TIMER2.handle_interrupt(),
-                        peripheral_interrupts::UART0 => uart::UARTE0.handle_interrupt(),
-                        peripheral_interrupts::SPI0_TWI0 => {
-                            // SPI0 and TWI0 share interrupts.
-                            // Dispatch the correct handler.
-                            match (spi::SPIM0.is_enabled(), i2c::TWIM0.is_enabled()) {
-                                (false, false) => (),
-                                (true, false) => spi::SPIM0.handle_interrupt(),
-                                (false, true) => i2c::TWIM0.handle_interrupt(),
-                                (true, true) => debug_assert!(
-                                    false,
-                                    "SPIM0 and TWIM0 cannot be \
-                                     enabled at the same time."
-                                ),
-                            }
-                        }
-                        peripheral_interrupts::SPI1_TWI1 => {
-                            // SPI1 and TWI1 share interrupts.
-                            // Dispatch the correct handler.
-                            match (spi::SPIM1.is_enabled(), i2c::TWIM1.is_enabled()) {
-                                (false, false) => (),
-                                (true, false) => spi::SPIM1.handle_interrupt(),
-                                (false, true) => i2c::TWIM1.handle_interrupt(),
-                                (true, true) => debug_assert!(
-                                    false,
-                                    "SPIM1 and TWIM1 cannot be \
-                                     enabled at the same time."
-                                ),
-                            }
-                        }
-                        peripheral_interrupts::SPIM2_SPIS2_SPI2 => spi::SPIM2.handle_interrupt(),
-                        peripheral_interrupts::ADC => adc::ADC.handle_interrupt(),
-                        _ => debug!("NvicIdx not supported by Tock"),
+                    if !self.interrupt_service.service_interrupt(interrupt) {
+                        debug!("NvicIdx not supported by Tock: {}", interrupt);
                     }
```

Comment:
What if this first calls `self.interrupt_service.service_interrupt(interrupt)` but then falls back to calling its local interrupt service function if `false` is returned? That way the other more specific crates only have to worry about the interrupts they care about and don't have to call the generic interrupt service code. In that case, I also think it would then make sense to make `self.interrupt_service` an `Option<&'static dyn InterruptServiceTrait>`.

Reply:
I find it more logical the other way around (i.e. as currently implemented).

If say there is a chain of more specific chip families, such as `nrf5x -> nrf52x -> nrf52840`, then the more specific ones depend on the more generic ones and the chain looks like this:

```rust
pub struct Nrf52840InterruptService {
    nrf52: nrf52::chip::Nrf52InterruptService,
    // other fields
}
pub struct Nrf52InterruptService {
    nrf5x: nrf5x::chip::Nrf5xInterruptService,
    // other fields
}
pub struct Nrf5xInterruptService {
    // other fields
}

impl InterruptService for Nrf52840InterruptService {
    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
        match interrupt {
            FOO_INTERRUPT => FOO.handle_interrupt(),
            _ => return self.nrf52.service_interrupt(interrupt),
        }
        true
    }
}
impl InterruptService for Nrf52InterruptService {
    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
        match interrupt {
            BAR_INTERRUPT => BAR.handle_interrupt(),
            _ => return self.nrf5x.service_interrupt(interrupt),
        }
        true
    }
}
impl InterruptService for Nrf5xInterruptService {
    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
        match interrupt {
            XYZ_INTERRUPT => XYZ.handle_interrupt(),
            _ => return false,
        }
        true
    }
}
```

If I understood correctly, your proposal would look more like this:

```rust
pub struct Nrf52840InterruptService {
    child_chip: Option<&dyn InterruptService>,
    // other fields
}
pub struct Nrf52InterruptService {
    child_chip: Option<&dyn InterruptService>,
    // other fields
}
pub struct Nrf5xInterruptService {
    child_chip: Option<&dyn InterruptService>,
    // other fields
}

impl InterruptService for Nrf52840InterruptService {
    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
        match interrupt {
            FOO_INTERRUPT => FOO.handle_interrupt(),
            _ => return self.child_chip.map(|service| service.service_interrupt(interrupt)),
        }
        true
    }
}
impl InterruptService for Nrf52InterruptService {
    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
        match interrupt {
            BAR_INTERRUPT => BAR.handle_interrupt(),
            _ => return self.child_chip.map(|service| service.service_interrupt(interrupt)),
        }
        true
    }
}
impl InterruptService for Nrf5xInterruptService {
    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
        match interrupt {
            XYZ_INTERRUPT => XYZ.handle_interrupt(),
            _ => return self.child_chip.map(|service| service.service_interrupt(interrupt)),
        }
        true
    }
}
```

I don't think the overhead in terms of an `Option`, a `dyn` and more code is worth it when it is known at compile-time that `nrf52840` always depends on `nrf52` which always depends on `nrf5x`.
Or do you have something else in mind that could work with such chains of chips?

Reply:
I had imagined it like this:
```rust
pub struct Nrf52840InterruptService {
    child_chip: Option<&dyn InterruptService>,
    // other fields
}
pub struct Nrf52InterruptService {
    child_chip: Option<&dyn InterruptService>,
    // other fields
}
pub struct Nrf5xInterruptService {
    child_chip: Option<&dyn InterruptService>,
    // other fields
}

impl InterruptService for Nrf52840InterruptService {
    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
        let ret = self.child_chip.map_or(false, |service| service.service_interrupt(interrupt))
        if !ret {
            match interrupt {
                FOO_INTERRUPT => {FOO.handle_interrupt(); true},
                _ => false,
            }
        } else {
            true
        }
    }
}
impl InterruptService for Nrf52InterruptService {
    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
        let ret = self.child_chip.map_or(false, |service| service.service_interrupt(interrupt))
        if !ret {
            match interrupt {
                BAR_INTERRUPT => {BAR.handle_interrupt(); true},
                _ => false,
            }
        } else {
            true
        }
    }
}
impl InterruptService for Nrf5xInterruptService {
    unsafe fn service_interrupt(&self, interrupt: u32) -> bool {
        let ret = self.child_chip.map_or(false, |service| service.service_interrupt(interrupt))
        if !ret {
            match interrupt {
                XYZ_INTERRUPT => {XYZ.handle_interrupt(); true},
                _ => false,
            }
        } else {
            true
        }
    }
}
```
and the reason is I think it creates a more straightforward pointer chain. The lowest level, `nrf5x` has a pointer to the `nrf5xIS`, which has a pointer to `nrf52IS`, which points to `nrf52840IS`. Then when an interrupt happens in the nrf5x it passes it up the chain. This I think better matches how callbacks propagate in the rest of Tock.

In the other approach the `nrf5x` has a pointer to the top of the chain (the nrf52840), and then calls down, where the second lowest layer (nrf52) actually calls back to the nrf5x, which is what started with the interrupt. That circle is not intuitive to me, and I find it unexpected that the nrf5x would rely on something else to call its own code.

However, in your approach the chain is more explicit in the code, which is nice, and I don't _think_ there would be any reason to alter the chain. Overall having the dependencies explicit is probably better.

Reply:
The top-down approach I suggest also has the advantage that the compiler could easily inline it all into a single match. With the reverse chain, there's some extra boilerplate and runtime overhead due to the dynamic typing, and the `Option`, even though in practice the types are static for a given chip. Probably not a huge runtime difference, but seems like an unnecessary overhead both on source code and on binary size.

And when you write there is a circle going back to nrf5x itself, I don't see it as a circle. It's more: `nrf5x::Chip -> nrf52840::IS -> nrf52::IS -> nrf5x::IS`. So it starts from the chip, but then goes to the chain of interrupt services - and the `nrf5x::IS` could be in a different module for that matter.
Ideally it could just be `nrf5x::Chip -> nrf52840::IS`, except that `nrf52840` defers some of its code to `nrf52`, not as part of a circle, but just to avoid code duplication.

To put things in perspective, a similar thing happens with the GPIOs for example: the `nrf5x` uses `gpio_port`, which is defined in `nrf52840`, but which then calls `nrf5x`'s GPIO code again. It's just a matter of organizing dependencies to avoid code duplication.

Reply:
I agree that logically the interrupt checking function called by the kernel in the nRF5x is different from the actual interrupt servicing code, but if I'm new and trying to understand how interrupts are serviced for the nRF52840 chip, I wouldn't expect that the code would end up looping around back to the same file. This is also why I like your approach that makes the types explicit so that following the chain does not require looking at the setup code.